### PR TITLE
[compiler] Custom type definitions in config

### PR DESCRIPTION
**Pull request by mvitousek**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #30670

Summary:
This PR allows custom type definitions for globals to be loaded in the Forget config. We use a slightly different representation than we do internally, in order to ensure that things are well formed before we pass them to the shape registry. The most fundamental issue is that we introduce a notion of IDs (expected to be unique) for types: we can introduce a type and then refer to that type later by a unique ID in a function return type if we have, for example, a global value that returns itself when called.

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/HIR/Environment.ts [hunk 1]
```diff
@@ -756,6 +847,7 @@ export class Environment {
     return (
       moduleName.toLowerCase() === 'react' ||
       moduleName.toLowerCase() === 'react-dom' ||
+      [...this.#globals.keys()].includes(moduleName) ||
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/Environment.ts:850]: why do we need this?

File: compiler/packages/babel-plugin-react-compiler/src/HIR/Globals.ts [hunk 2]
```diff
@@ -516,6 +519,68 @@ DEFAULT_GLOBALS.set(
...
+        func.returnType = {kind: 'Object', shapeId: BuiltInArrayId};
+      } else if (refMap.has(type.fn.returnType)) {
+        func.returnType = refMap.get(type.fn.returnType)!;
+      } else {
+        const cur = toSet.get(type.fn.returnType);
+        toSet.set(type.fn.returnType, (ty: BuiltInType) => {
+          cur?.(ty);
+          func.returnType = ty;
+        });
+      }
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/Globals.ts:558]: does this allow custom types to refer to builtin types? would be nice if that was the case (helper function that returns a ref, for example)

File: compiler/packages/babel-plugin-react-compiler/src/HIR/Environment.ts (hunk 1 above)
Comment (mvitousek) [reply compiler/packages/babel-plugin-react-compiler/src/HIR/Environment.ts:850]: This predicate gates giving types to imports from globals [here](https://github.com/facebook/react/pull/30670/files/6be688d88d66e16701d5fab08c364d40e4dcbdbb#diff-3ebfb2fef1f4c7802f493c87477cce82d5a1385b5c39a887db981a4308e2e23dR813-R816). Gonna rename this function to `isKnownTypedModule`!

File: compiler/packages/babel-plugin-react-compiler/src/HIR/Globals.ts (hunk 2 above)
Comment (mvitousek) [reply compiler/packages/babel-plugin-react-compiler/src/HIR/Globals.ts:558]: Yeah! Just need to add 'ref', etc to `TypeReferenceSchema` and then link it with the builtin ref shape.

**Review by josephsavona — APPROVED** [review]
nice! thanks for confirming how that works, and i guess we can always add more cases for the built-in types as we need them

**Review by josephsavona — CHANGES_REQUESTED** [review]
As part of internal rollout I needed more flexibility than this (ability to supply a type for arbitrary modules of a certain pattern) so I landed #30771. Let's chat when you're back, i'm happy to iterate on the design (i borrowed ideas from this PR)

**Pull request closed without merging**